[TensorExpr] Add aten::matmuls to TE fuser.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54605 [TensorExpr] Add aten::matmuls to TE fuser.**

For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

Commit:
[TensorExpr] Add aten::matmuls to TE fuser.

For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

eellison reviewed (commented):
You should be aware that matmul has broadcasting behavior, so either make sure that's handled or opt out in fuser

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

ZolotukhinM commented:
> You should be aware that matmul has broadcasting behavior, so either make sure that's handled or opt out in fuser

That's a good point, I totally forgot about that! I've added checks into the fuser and tests to verify that we don't fuse what we can't correctly lower yet. Could you please take a look? Are these tests sufficient?

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -187,6 +187,26 @@ bool conv2dIsSupported(const torch::jit::Node* node) {
 }
 
+// The fuser currently only supports matmul of 2D x 2D matrices
+bool matmulIsSupported(const torch::jit::Node* node) {
```
Is there a reason for putting this in kernel instead of tensorexpr_fuser ?

ZolotukhinM replied:
Only because we already have `conv2dIsSupported` there. I wouldn't mind moving it to `tensorexpr_fuser.cpp`, but in this PR for consistency I'd prefer to keep them in `kernel.cpp`

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2060,6 +2084,42 @@ Tensor* TensorExprKernel::computeConv2d(const torch::jit::Value* v) {
+  const IntImm* total_size = dynamic_cast<const IntImm*>(
+      IRSimplifier::simplify((size_a[0] * size_a[1] * size_b[1])).node());
+
+  if (total_size && total_size->value() < 1000) {
```
Can you add a comment here about why we're using external calls for n > 1000 ?

ZolotukhinM replied:
👍

bertmaher replied:
lol clang-tidy "1000 is a magic number".  I want to troll the linter by doing `constexpr int kMagicNumber = 1000;`

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2060,6 +2084,42 @@ Tensor* TensorExprKernel::computeConv2d(const torch::jit::Value* v) {
+  const Node* n = v->node();
+  auto const& shape = sizesForValue(v);
+  Dtype dtype = kFloat;
+  auto maybe_stype = findDtypeForValue(v);
```
What is maybe_stype?

ZolotukhinM replied:
"maybe" is for optionality of returned value, "stype" is for scalar-type. I think we've been using this name in surrounding code too.

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2060,6 +2084,42 @@ Tensor* TensorExprKernel::computeConv2d(const torch::jit::Value* v) {
+  auto size_a = ExprVectorToExprHandleVector(a->dims());
+  auto size_b = ExprVectorToExprHandleVector(b->dims());
+  const IntImm* total_size = dynamic_cast<const IntImm*>(
+      IRSimplifier::simplify((size_a[0] * size_a[1] * size_b[1])).node());
```
Should we be multiplying by `size_b[0]` as well ?

ZolotukhinM replied:
It's a rough estimate of amount of work matmul needs to do, which is N*M*K (size_a[1] == size_b[0] == M). Would I not pass an interview with such analysis? 😜

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2060,6 +2084,42 @@ Tensor* TensorExprKernel::computeConv2d(const torch::jit::Value* v) {
+          BufHandle bh(b);
+          return Load::make(ah, {m, k}) * Load::make(bh, {k, n});
+        },
+        {{size_a[1], "K"}});
```
In the spirit of https://randomascii.wordpress.com/2014/01/27/theres-only-four-billion-floatsso-test-them-all/: there are a very small number of possible inputs to test, you might try writing a script to see how large you can make the total size while nnc is as fast as aten matmul (with a little bit of a leeway for possible fusion, say 5%)

ZolotukhinM replied:
This is a very basic naive heuristic and we for sure would like to tune it (and hopefully add better schedules for matmul as well). But I think it deserves a separate PR too, this one is mostly to lay down the foundation for this work.

eellison reviewed (commented):
LGTM, just had one question about `b[0]`

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add aten::matmuls to TE fuser."


For small sizes we generate a naive 3-layer loopnest, for bigger sizes
we generate an external call.

Differential Revision: [D27298364](https://our.internmc.facebook.com/intern/diff/D27298364)

[ghstack-poisoned]

bertmaher reviewed (approved):
Looks good to me.  So if we fuse matmul and use an external call, are we still as fast as eager mode?

ZolotukhinM commented:
> So if we fuse matmul and use an external call, are we still as fast as eager mode?

I would expect so, but honestly I haven't done any rigorous measurements yet.
